File: packages/sim/src/terminals/generators/phishing.ts
import type { Rng } from "../../core/rng";
import type { MessageContent } from "../../world/world";
import type { ClassifyQuestion } from "../types";
import { base, fill, param, pool, type GenContext } from "./common";

interface MessageDef {
  id: string;
  from: string;
  subject: string;
  body: string;
  signals: string[];
  explanation: string;
}

interface PhishingPool {
  labels: { from: string; signalsTitle: string; explanationTemplate: string };
  messages: { phish: MessageDef[]; legit: MessageDef[] };
  classify: { prompt: string; categories: string[]; explanation: string; hint: string };
}

/** Conteúdo de uma mensagem do HUD (isca de phishing ou comunicado legítimo). */
export function makeInboxMessage(rng: Rng, kind: "phish" | "legit", pools: Record<string, unknown>): MessageContent {
  const p = pools.phishing as PhishingPool | undefined;
  if (!p) throw new Error("pool phishing ausente");
  const m = rng.pick(p.messages[kind]);
  return { from: m.from, subject: m.subject, body: m.body, signals: [...m.signals] };
}

/**
 * Triagem de caixa de entrada: metade das mensagens é fraude.
 * Os itens mostram só o que o jogador veria (remetente, assunto, corpo);
 * os sinais ficam na explicação de cada item.
 */
export function phishingClassify(ctx: GenContext): ClassifyQuestion {
  const P = pool<PhishingPool>(ctx, "phishing");
  const rng = ctx.rng;
  const count = param(ctx, "count", 4);
  const nPhish = Math.min(Math.ceil(count / 2), P.messages.phish.length);
  const nLegit = Math.min(count - nPhish, P.messages.legit.length);

  const built = [
    ...rng.sample(P.messages.phish, nPhish).map((m) => ({ m, category: 0 })),
    ...rng.sample(P.messages.legit, nLegit).map((m) => ({ m, category: 1 })),
  ];
  const chosen = rng.shuffle(built.map((_, i) => i)).map((i) => built[i]!);

  return {
    ...base(ctx, "phishing_classify"),
    kind: "classify",
    prompt: P.classify.prompt,
    categories: P.classify.categories,
    items: chosen.map(({ m }) => ({
      text: m.subject,
      detail: [fill(P.labels.from, { from: m.from }), m.body],
    })),
    answer: chosen.map((c) => c.category),
    itemExplanations: chosen.map(({ m }) =>
      fill(P.labels.explanationTemplate, { explanation: m.explanation, signals: m.signals.join("; ") }),
    ),
    explanation: P.classify.explanation,
    hint: P.classify.hint,
    tags: ["phishing_items"],
  };
}

File: apps/api/app/terminals/generators/phishing.py
"""Triagem de phishing — espelho de packages/sim/src/terminals/generators/phishing.ts.

`makeInboxMessage` (mensagens do HUD) não tem espelho: é conteúdo de jogo, não questão.
"""

from __future__ import annotations

import math
from typing import Any

from .common import GenContext, base, fill, param, pool


def phishing_classify(ctx: GenContext) -> dict[str, Any]:
    p = pool(ctx, "phishing")
    rng = ctx.rng
    labels = p["labels"]
    classify = p["classify"]
    count = param(ctx, "count", 4)
    phish: list[dict[str, Any]] = p["messages"]["phish"]
    legit: list[dict[str, Any]] = p["messages"]["legit"]
    n_phish = min(math.ceil(count / 2), len(phish))
    n_legit = min(count - n_phish, len(legit))

    built = [{"m": m, "category": 0} for m in rng.sample(phish, n_phish)]
    built += [{"m": m, "category": 1} for m in rng.sample(legit, n_legit)]
    chosen = [built[i] for i in rng.shuffle(list(range(len(built))))]

    return {
        **base(ctx, "phishing_classify"),
        "kind": "classify",
        "prompt": classify["prompt"],
        "categories": classify["categories"],
        "items": [
            {
                "text": c["m"]["subject"],
                "detail": [fill(labels["from"], {"from": c["m"]["from"]}), c["m"]["body"]],
            }
            for c in chosen
        ],
        "answer": [c["category"] for c in chosen],
        "itemExplanations": [
            fill(
                labels["explanationTemplate"],
                {"explanation": c["m"]["explanation"], "signals": "; ".join(c["m"]["signals"])},
            )
            for c in chosen
        ],
        "explanation": classify["explanation"],
        "hint": classify["hint"],
        "tags": ["phishing_items"],
    }
